R/utils-cut.R

```diff
@@ -43,26 +43,107 @@ cut_width <- function(x, width, center = NULL, boundary = NULL, closed = c("righ
+  # Perhaps we can rationalise/remove the boundary code if a use-case can't
+  # be described.
   min_x <- find_origin(x_range, width, boundary)
   max_x <- max(x, na.rm = TRUE)
 
   breaks <- c(seq(min_x, max_x, width))
+  # add an extra break if the seq() didn't produce a break >= max_x
+  if (tail(breaks, 1) < max_x) {
+    breaks <- c(breaks, tail(breaks, 1) + width)
+  }
+
+  # safety checks
+  if (tail(breaks, 1) < max_x) {
+    stop("Problem with final bin in utils-cut.R")
+  }
+  if (length(breaks) < 2) {
+    stop("Less than two breaks in utils-cut.R")
+  }
+
+
   # Round breaks *before* they go into the cut function. This way the data (not rounded)
   # will be correctly divided into the rounded bins
-  if (all(x %% 1 == 0)) {
-    # if all integers, use default formatting (6 significant digits)
+  if (all(breaks %% 1 == 0)) {
+    # if breaks are all integers, skip our own rounding
+    # and use default formatting in cut() (6 significant digits)
     requiredDigits <- -1
   } else {
-    requiredDigits <- ceiling(median(stringi::stri_count_regex(as.character(x), "[[:digit:]]")))
+    # Determine the appropriate number of digits to round the bin start/ends to.
+    # Start at 1 and work up!
+    requiredDigits <- 0
+    rawBreaks <- breaks
+    repeat {
+      requiredDigits <- requiredDigits + 1
+      if (requiredDigits > 20) { # safety escape condition
+        break
+      }
+
+      # round the breaks to requiredDigits significant digits
+      # we use formatC (not signif) because the cut() below uses it too,
+      # and we are aiming for consistency (though maybe we could pass our labels to it?)
+      breaks <- as.numeric(formatC(0 + rawBreaks, digits = requiredDigits, width = 1L))
+
+      # if any breaks are the same, try more significant digits
+      if (anyDuplicated(breaks) > 0) {
+        next
+      }
+
+      # calculate the inter-break differences (binWidths) from rounded values
+      diffs <- diff(breaks)
+
+      # if only one difference (e.g. one bin), we can't do the following part
+      # which checks that the differences are not too different from each other
+      if (length(diffs) < 2) {
+        break
+      }
+
+      # now see if the normalised variance (index of dispersion) of the diffs is small enough to
+      # stop increasing the number of significant digits
+      dispersion <- var(diffs)/mean(diffs)
+      if (dispersion < 1e-05) {
+        break
+      }
+    }
+  }
+
+  # Now, we know we have enough bins to encompass min_x and max_x, but they could have been
+  # rounded to slightly just above or below, respectively.
+  # If so, we will adjust the first and last bins by subtracting/adding one "least significant digit"
+  # from them.
+
+## DEBUG ONLY
+unmodifiedBreaks <- breaks
+
+  if (min_x < breaks[1]) {
+    newStart <- breaks[1] - signifDigitDelta(breaks[1], requiredDigits)
+    breaks[1] <- as.numeric(formatC(newStart, digits = requiredDigits, width = 1L))
+    print("MODIFIED START BIN")
```

Comment:
found it! let's use `veupathUtils::logWithTime('blah', verbose)` and i think just passing in `verbose` as an arg with default to TRUE should do it!

Reply:
Cool - but I don't think these tiny modifications are worth logging.  I pushed the changes in a hurry so you could take a look. I don't think these should stay in the final push.